File: src/domain/journal.ts
import { LIMITS } from './limits';
import { Timeline, dailyRestStatus, isWeeklyRest, weeklyRestStatus } from './shifts';
import { MINUTE, WEEK, minutesBetween, overlapMinutes, weekStartUtc } from './time';
import type { CrewMode, ManualShift, RestKind, RestStatus, ShiftMeta } from './types';

export type Level = 'ok' | 'warn' | 'bad';

export interface JournalShift {
  id: string;
  source: 'auto' | 'manual';
  start: number;
  end: number | null;
  startCountry: string | null;
  endCountry: string | null;
  driveMinutes: number;
  workMinutes: number;
  poaMinutes: number;
  breakMinutes: number;
  spanMinutes: number;
  continuousDriveAtEndMinutes: number;
  rest: {
    kind: RestKind;
    minutes: number;
    ongoing: boolean;
    split: boolean;
    status: RestStatus | null;
  };
  notes: string;
  /**
   * Смена «живая»: идёт сейчас или после неё идёт отдых. Её правки меняют
   * записи режимов, остальные смены можно сохранить как ручные.
   */
  live: boolean;
  /** Конец отдыха после смены (начало следующей), если он завершён. */
  restEnd: number | null;
  /** Подсветка по дизайну: 10 ч вождения, 13+ ч смены, сокращённый отдых. */
  levels: { drive: Level; span: Level; rest: Level };
}

export interface JournalWeek {
  start: number;
  isCurrent: boolean;
  driveMinutes: number;
  fortnightMinutes: number;
  shifts: JournalShift[];
  weeklyRests: { start: number; end: number | null; minutes: number; status: RestStatus }[];
}

interface Options {
  timeline: Timeline;
  manualShifts: ManualShift[];
  meta: Record<string, ShiftMeta>;
  crewMode: CrewMode;
  now: number;
}

export function buildJournal({ timeline, manualShifts, meta, crewMode, now }: Options): JournalWeek[] {
  const team = crewMode === 'TEAM';
  const spanLevel = (m: number): Level => {
    if (team) return m > LIMITS.shiftWindowTeam - LIMITS.teamDailyRest ? 'bad' : 'ok';
    if (m > LIMITS.shiftWindowSolo - LIMITS.dailyRestReduced) return 'bad';
    return m > LIMITS.shiftWindowSolo - LIMITS.dailyRestRegular ? 'warn' : 'ok';
  };
  const driveLevel = (m: number): Level =>
    m > LIMITS.dailyDriveExtended ? 'bad' : m > LIMITS.dailyDrive ? 'warn' : 'ok';
  const restLevel = (s: RestStatus | null): Level => (s === 'insufficient' ? 'bad' : s === 'reduced' ? 'warn' : 'ok');

  const shifts: JournalShift[] = [];

  for (const s of timeline.shifts) {
    const m = meta[s.id] ?? {};
    const r = s.restAfter;
    const kind: RestKind = r ? (isWeeklyRest(r) ? 'weekly' : 'daily') : 'none';
    const status = !r || r.open ? null : kind === 'weekly' ? weeklyRestStatus(r.restMinutes) : dailyRestStatus(r.restMinutes, s.splitFirstPart);
    const spanMinutes = minutesBetween(s.start, s.end ?? now);
    shifts.push({
      id: s.id,
      source: 'auto',
      start: s.start,
      end: s.end,
      startCountry: m.startCountry ?? null,
      endCountry: m.endCountry ?? null,
      driveMinutes: s.driveMinutes,
      workMinutes: s.workMinutes,
      poaMinutes: s.poaMinutes,
      breakMinutes: s.breakMinutes,
      spanMinutes,
      continuousDriveAtEndMinutes: s.continuousDriveAtEnd,
      rest: { kind, minutes: r?.restMinutes ?? 0, ongoing: r?.open ?? false, split: s.splitFirstPart, status },
      notes: m.notes ?? '',
      live: s.end === null || (r?.open ?? false),
      restEnd: r && !r.open ? r.end : null,
      levels: { drive: driveLevel(s.driveMinutes), span: spanLevel(spanMinutes), rest: restLevel(status) },
    });
  }

  for (const ms of manualShifts) {
    const spanMinutes = minutesBetween(ms.start, ms.end ?? now);
    const status =
      ms.rest.kind === 'daily'
        ? dailyRestStatus(ms.rest.minutes, ms.rest.split)
        : ms.rest.kind === 'weekly'
          ? weeklyRestStatus(ms.rest.minutes)
          : null;
    shifts.push({
      id: ms.id,
      source: 'manual',
      start: ms.start,
      end: ms.end,
      startCountry: ms.startCountry,
      endCountry: ms.endCountry,
      driveMinutes: ms.driveMinutes,
      workMinutes: 0,
      poaMinutes: 0,
      breakMinutes: 0,
      spanMinutes,
      continuousDriveAtEndMinutes: ms.continuousDriveAtEndMinutes,
      rest: { kind: ms.rest.kind, minutes: ms.rest.minutes, ongoing: false, split: ms.rest.split, status },
      notes: ms.notes,
      live: false,
      restEnd: ms.end !== null && ms.rest.kind !== 'none' ? ms.end + ms.rest.minutes * MINUTE : null,
      levels: { drive: driveLevel(ms.driveMinutes), span: spanLevel(spanMinutes), rest: restLevel(status) },
    });
  }

  // Недели — по началу смены; вождение недели — точно по отрезкам
  const currentWeek = weekStartUtc(now);
  const weekStarts = new Set<number>([currentWeek, ...shifts.map((s) => weekStartUtc(s.start))]);
  const driveInWeek = (from: number) => {
    let total = 0;
    for (const b of timeline.blocks) {
      if (b.activity === 'DRIVE') total += overlapMinutes(b.start, b.end, from, from + WEEK);
    }
    for (const ms of manualShifts) if (ms.start >= from && ms.start < from + WEEK) total += ms.driveMinutes;
    return total;
  };

  const weeklyRests = [
    ...timeline.rests.filter(isWeeklyRest).map((p) => ({
      start: p.start,
      end: p.open ? null : p.end,
      minutes: p.restMinutes,
      status: weeklyRestStatus(p.restMinutes),
    })),
    ...manualShifts
      .filter((ms) => ms.rest.kind === 'weekly' && ms.end !== null)
      // Тот же отдых уже есть в записях режимов — не показываем его дважды
      .filter(
        (ms) =>
          !timeline.rests.some(
            (p) => isWeeklyRest(p) && p.start < ms.end! + ms.rest.minutes * MINUTE && (p.open ? now : p.end) > ms.end!,
          ),
      )
      .map((ms) => ({
        start: ms.end!,
        end: ms.end! + ms.rest.minutes * MINUTE,
        minutes: ms.rest.minutes,
        status: weeklyRestStatus(ms.rest.minutes),
      })),
  ];

  return [...weekStarts]
    .sort((a, b) => b - a)
    .map((start) => {
      const driveMinutes = driveInWeek(start);
      return {
        start,
        isCurrent: start === currentWeek,
        driveMinutes,
        fortnightMinutes: driveMinutes + driveInWeek(start - WEEK),
        shifts: shifts.filter((s) => weekStartUtc(s.start) === start).sort((a, b) => b.start - a.start),
        // Недельный отдых показываем в неделе, где он закончился (или идёт)
        weeklyRests: weeklyRests.filter((r) => weekStartUtc(r.end ?? now) === start),
      };
    });
}

File: packages/tacho_engine/lib/src/journal.dart
import 'package:meta/meta.dart';
import 'package:tacho_engine/src/compliance.dart';
import 'package:tacho_engine/src/driver_mode.dart';
import 'package:tacho_engine/src/eu_limits.dart';
import 'package:tacho_engine/src/manual_shift.dart';
import 'package:tacho_engine/src/shifts.dart';
import 'package:tacho_engine/src/time.dart';

/// Подсветка значения в журнале по дизайну.
enum JournalLevel { ok, warn, bad }

/// Отдых после смены в журнале.
@immutable
class JournalRest {
  const new({
    required this.kind,
    required this.duration,
    required this.ongoing,
    required this.split,
    required this.status,
  });

  final RestKind kind;
  final Duration duration;
  final bool ongoing;

  /// Раздельный отдых 3 + 9.
  final bool split;

  /// null — отдыха нет или он ещё идёт.
  final RestStatus? status;
}

/// Смена в журнале: из записей режимов или внесённая вручную.
@immutable
class JournalShift {
  const new({
    required this.recorded,
    required this.manual,
    required this.start,
    required this.end,
    required this.driving,
    required this.otherWork,
    required this.availability,
    required this.breaks,
    required this.span,
    required this.continuousDrivingAtEnd,
    required this.rest,
    required this.live,
    required this.restEnd,
    required this.driveLevel,
    required this.spanLevel,
    required this.restLevel,
  });

  /// Смена из записей режимов; null — ручная.
  final Shift? recorded;

  /// Смена, внесённая итогами; null — из записей.
  final ManualShift? manual;

  final DateTime start;
  final DateTime? end;
  final Duration driving;
  final Duration otherWork;
  final Duration availability;
  final Duration breaks;

  /// Рабочий день: от начала смены до её конца (или до сейчас).
  final Duration span;
  final Duration continuousDrivingAtEnd;
  final JournalRest rest;

  /// Смена «живая»: идёт сейчас или после неё идёт отдых. Её правки меняют
  /// записи режимов, остальные смены можно сохранить как ручные.
  final bool live;

  /// Конец отдыха после смены (начало следующей), если он завершён.
  final DateTime? restEnd;

  /// 10 ч вождения — плохо, больше 9 ч — продление.
  final JournalLevel driveLevel;

  /// Больше 15 ч — плохо, больше 13 ч — только при сокращённом отдыхе.
  final JournalLevel spanLevel;
  final JournalLevel restLevel;
}

/// Неделя журнала: с понедельника 00:00 UTC.
@immutable
class JournalWeek {
  const new({
    required this.start,
    required this.isCurrent,
    required this.driving,
    required this.fortnightDriving,
    required this.shifts,
    required this.weeklyRests,
  });

  final DateTime start;
  final bool isCurrent;

  /// Вождение за неделю — точно по отрезкам, даже если смена переходит
  /// через полночь понедельника.
  final Duration driving;

  /// Вождение за эту и прошлую неделю (лимит 90 ч).
  final Duration fortnightDriving;

  /// Смены, начавшиеся на этой неделе, от новых к старым.
  final List<JournalShift> shifts;

  /// Недельные отдыхи, закончившиеся (или идущие) на этой неделе.
  final List<WeeklyRest> weeklyRests;
}

/// Журнал по неделям, от текущей к старым. Недели без смен не выводятся,
/// кроме текущей.
List<JournalWeek> buildJournal({
  required Timeline timeline,
  required DateTime now,
  Iterable<ManualShift> manualShifts = const [],
  CrewMode crew = CrewMode.solo,
}) {
  final manual = manualShifts.toList();
  final team = crew == CrewMode.team;

  JournalLevel spanLevel(Duration span) {
    if (team) {
      final limit = EuLimits.teamWorkdayWindow - EuLimits.teamDailyRest;
      return span > limit ? JournalLevel.bad : JournalLevel.ok;
    }
    if (span > EuLimits.workdayWithReducedRest) return JournalLevel.bad;
    return span > EuLimits.workdayWithRegularRest
        ? JournalLevel.warn
        : JournalLevel.ok;
  }

  JournalLevel driveLevel(Duration driving) {
    if (driving > EuLimits.dailyDrivingExtended) return JournalLevel.bad;
    return driving > EuLimits.dailyDriving
        ? JournalLevel.warn
        : JournalLevel.ok;
  }

  JournalLevel restLevel(RestStatus? status) => switch (status) {
    RestStatus.insufficient => JournalLevel.bad,
    RestStatus.reduced => JournalLevel.warn,
    RestStatus.full || null => JournalLevel.ok,
  };

  final shifts = <JournalShift>[];

  for (final s in timeline.shifts) {
    final r = s.restAfter;
    final kind = r == null
        ? RestKind.none
        : (r.isWeekly ? RestKind.weekly : RestKind.daily);
    final status = r == null || r.open
        ? null
        : kind == RestKind.weekly
        ? weeklyRestStatus(r.rest)
        : dailyRestStatus(
            shiftRestInWindow(timeline.blocks, s, crew),
            split: s.splitFirstPart,
          );
    final span = durationBetween(s.start, s.end ?? now);
    shifts.add(
      JournalShift(
        recorded: s,
        manual: null,
        start: s.start,
        end: s.end,
        driving: s.driving,
        otherWork: s.otherWork,
        availability: s.availability,
        breaks: s.breaks,
        span: span,
        continuousDrivingAtEnd: s.continuousDrivingAtEnd,
        rest: JournalRest(
          kind: kind,
          duration: r?.rest ?? Duration.zero,
          ongoing: r?.open ?? false,
          split: s.splitFirstPart,
          status: status,
        ),
        live: s.isOngoing || (r?.open ?? false),
        restEnd: r != null && !r.open ? r.end : null,
        driveLevel: driveLevel(s.driving),
        spanLevel: spanLevel(span),
        restLevel: restLevel(status),
      ),
    );
  }

  for (final m in manual) {
    final span = durationBetween(m.start, m.end ?? now);
    final status = switch (m.restKind) {
      RestKind.daily => dailyRestStatus(
        restInWindow(span, m.rest, crew),
        split: m.splitRest,
      ),
      RestKind.weekly => weeklyRestStatus(m.rest),
      RestKind.none => null,
    };
    shifts.add(
      JournalShift(
        recorded: null,
        manual: m,
        start: m.start,
        end: m.end,
        driving: m.driving,
        otherWork: Duration.zero,
        availability: Duration.zero,
        breaks: Duration.zero,
        span: span,
        continuousDrivingAtEnd: m.continuousDrivingAtEnd,
        rest: JournalRest(
          kind: m.restKind,
          duration: m.rest,
          ongoing: false,
          split: m.splitRest,
          status: status,
        ),
        live: false,
        restEnd: m.restEnd,
        driveLevel: driveLevel(m.driving),
        spanLevel: spanLevel(span),
        restLevel: restLevel(status),
      ),
    );
  }

  // Недели — по началу смены; вождение недели — точно по отрезкам
  final currentWeek = weekStartUtc(now);
  final weekStarts = {
    currentWeek,
    for (final s in shifts) weekStartUtc(s.start),
  };
  Duration drivingInWeek(DateTime from) {
    final to = from.add(week);
    var total = Duration.zero;
    for (final b in timeline.blocks) {
      if (b.mode == DriverMode.driving) {
        total += overlap(b.start, b.end, from, to);
      }
    }
    for (final m in manual) {
      if (!m.start.isBefore(from) && m.start.isBefore(to)) total += m.driving;
    }
    return total;
  }

  final recordedWeekly = timeline.rests.where((p) => p.isWeekly).toList();
  final weeklyRests = [
    for (final p in recordedWeekly)
      WeeklyRest(
        start: p.start,
        end: p.open ? null : p.end,
        duration: p.rest,
        status: weeklyRestStatus(p.rest),
      ),
    for (final m in manual)
      if (m.restKind == RestKind.weekly && m.end != null)
        // Тот же отдых уже есть в записях режимов — не показываем его дважды
        if (!recordedWeekly.any(
          (p) =>
              p.start.isBefore(m.restEnd!) &&
              (p.open ? now : p.end).isAfter(m.end!),
        ))
          WeeklyRest(
            start: m.end!,
            end: m.restEnd,
            duration: m.rest,
            status: weeklyRestStatus(m.rest),
          ),
  ];

  final sortedWeeks = weekStarts.toList()..sort((a, b) => b.compareTo(a));
  return [
    for (final start in sortedWeeks)
      JournalWeek(
        start: start,
        isCurrent: start == currentWeek,
        driving: drivingInWeek(start),
        fortnightDriving:
            drivingInWeek(start) + drivingInWeek(start.subtract(week)),
        shifts: shifts.where((s) => weekStartUtc(s.start) == start).toList()
          ..sort((a, b) => b.start.compareTo(a.start)),
        // Недельный отдых показываем в неделе, где он закончился (или идёт)
        weeklyRests: weeklyRests
            .where((r) => weekStartUtc(r.end ?? now) == start)
            .toList(),
      ),
  ];
}
